posts/copilot-workspace/index.qmd

```diff
@@ -0,0 +1,102 @@
+
+## Experiment 3: Conducting a code review on an R package
+
+::: {.callout-note}
+Difficulty level: moderate to high
+:::
+
+In the last test case, we requested GitHub copilot workspace to conduct a [code review of the entire code base of a package](https://github.com/epiverse-trace/simulist/pull/117). At Epiverse-TRACE, it is standard practise to conduct a full package review. As any code base develops, including R packages, more commits are added and often the number of lines of code increases. Reviewing all of these changes before releasing a new version of the package is a timely undertaking for package reviewers. If AI tools could review the code base, similar to how static code analysers work, and suggest improvements to, for example, code style, memory efficiency, documentation, unit testing, UX/UI, it would greatly improve efficiency before releasing the code.
+
+### Outcome
+
+Here, Copilot provided no analysis of the code base. It regurgitated the changes described in the [package review pull request](https://epiverse-trace.github.io/blueprints/code-review.html#full-package-review), likely from a mix of the pull request description and the `NEWS.md` file, and did not analyse any of the code. Therefore, it provided no additional use than if the reviewer had just read the `NEWS`.
+
+## Conclusion
+
+The evidence above suggests there’s still a long way to go before AI-powered tools produce code projects unsupervised that are equivalent to that from a research software engineer. One issue is the generative AI has training database that leans towards common problems, which can makes its solutions to coding problems with specific constraints suboptimal, particularly in fields where there is a relatively small public code base to train models with (e.g. R functions for epidemic analysis are much rarer than generic JavaScript functions for web design). A second is that AI-models are picking up common patterns and synthesising them, and didn’t seem to be able to understand how different parts of a codebase link together, which a developer would be able to use to reduce duplication and keep a consistent design pattern. Put together, we think the applicability of current generative AI models to aid research software engineering in domain-specific R package development, where problems can be complex bespoke solutions are common, is likely to be low. This could improve for use-cases which seek to implement more common tasks, such as building dashboard templates.
```

Comment:
I think this paragraph requires further improving to clearly communicate the conclusions. I tried to reword it in the way that I understood it, hoping it did not change the original meaning.

- "The evidence above suggests there’s still a long way to go before AI-powered tools can produce unsupervised code projects equivalent to that of a research software engineer." - I don't think this should be the leading statement because it suggests that's what we set out to prove. I've moved another statement up here.

```suggestion
Based on our trials we observe that GitHub copilot Workspace may not provide a significant benefit to research software engineers in domain-specific R package development, where problems are complex and bespoke solutions are common.  The evidence above suggests there’s still a long way to go before AI-powered tools can produce unsupervised code projects equivalent to that of a research software engineer. One observation is that generative AI is better trained for some purposes than others, biased by the amount of training data/information available. This makes it struggle in niche areas with a relatively smaller public code base to train models with (e.g. R functions for epidemic analysis are much rarer than generic JavaScript functions for web design). A second is that AI models don’t seem to understand how different parts of a codebase link together, so they provide solutions that are inconsistent with the requirements of the codebase (design principles, code style, etc). These models may, however, be more useful for more common tasks, such as building dashboard templates.
```